### DISCUSSION TITLE
Native mode window displays but no content

### DISCUSSION THREAD

**Discussion by ed2050** [zauberzeug/nicegui#1674] [Q&A]
### Question

Hi, I'm getting started with nicegui.  Looks fantastic so far.  I can open a webpage interface no problem, using the [styling example](port=native_mode.find_open_port()).  Everything displays as it should.

When I switch to native mode, things fall apart.  A window opens.  Titlebar and window size are correct from params to ui.run.  But window content is completely empty.  Just a blank white canvas with no text or anything.

# Question
Can anyone point me to more info on this problem, or how to gather debug info?  I'm drawing a blank so far.

# What I tried
I'm running on mac.  I don't see any errors in the terminal.  All I get is this:
```
morbo> python3 nicegui-test.py 
NiceGUI ready to go on http://localhost:8000 (install netifaces to show all IPs and speedup this message)

# nothing more shows while script is running.  When I exit window, this appears :

/Library/Developer/CommandLineTools/Library/Frameworks/Python3.framework/Versions/3.8/lib/python3.8/multiprocessing/resource_tracker.py:216: UserWarning: resource_tracker: There appear to be 12 leaked semaphore objects to clean up at shutdown
  warnings.warn('resource_tracker: There appear to be %d '
```

I read the docs and discussion board, tried ui.page decorator, tried the port arg to ui.run, tried reload = false, nothing changes.  Native window is always blank.  I even ran all the code at top level of the script, not wrapped in main ().  Doesn't make a difference.

# Code
Here's my code.
```py
from nicegui import ui , native_mode

#@ui.page ('/')
def main () :

    ui.label ('Wassup')
    ui.icon('thumb_up')
    ui.markdown('This is **Markdown**.')
    ui.html('This is <strong>HTML</strong>.')
    
    with ui.row() :
        ui.label('CSS').style('background : #000; color: #888; font-weight: bold')
        ui.label('Tailwind').classes('font-serif')
        ui.label('Quasar').classes('q-ml-xl')
    ui.link('NiceGUI on GitHub', 'https://github.com/zauberzeug/nicegui')
    
    ui.run (title = 'My gui test', native = true ,
        port = native_mode.find_open_port () ,
        #reload = false,
        window_size = (900,900)
    )

if __name__ in { '__main__', '__mp_main__' } :
    main ()
```

# Environment
Here's my setup (revelant packages only):
```
macos catalina 10.15.7
python 3.8.2 (built in)
## pip packages
fastapi                   0.103.1     
fastapi-socketio          0.0.10      
nicegui                   1.3.14      
pywebview                 4.3.3 
streamlit                 1.26.0      
starlette                 0.27.0      
uvicorn                   0.22.0      
watchfiles                0.20.0      
websockets                11.0.3      
```

Oh one more thing.  Pywebview runs just fine.  I can open a native window with content, no problem.  For example this works perfectly fine:
```py
import webview
webview.create_window ('Test view', html = '<html><body><h1>Testing</h1><p>LeviOOOsa</p></body></html>')
webview.start ()
```

Any suggestions how to figure this out?

**Comment by rodja** [comment]
It's working for me after I changed `native = true` to `native = True`. What version of NiceGUI is installed? Best to verify in the code with `print(nicegui.__version__)`.

**Reply by ed2050** [reply]
Thanks Rodja.  Yeah I mapped true / false to True / False because uppercase primitives are silly. :) Forgot to change it before posting.

My `nicegui.__version__` shows same, 1.3.14.  I know it's very difficult to diagnose someone else's system remotely.  Hoping your experience can at least point me in the right direction.  Some ideas:

1) Is there a way to turn on debug / verbose output?  Sometimes logging messages get swallowed, especially if multiple threads or processes are involved.  Could be an error happening that I'm just not seeing in terminal.

2) Are there certain components in the stack I should look at that can cause these kinds of problems?  Or components I can test independently?  Not knowing the internal architecture I'm at a loss for where to look.  Never used Vue or Quasar, etc. 

3) Is there a test suite I can run to verify that everything in the package works?
Thanks.

**Reply by rodja** [reply]
1. There is no dedicated debug logging.
2. pywebview is working for you, but you could try to open the url from native mode in a browser. Does that work?
3. Yes, NiceGUI has extensive tests for its UI elements. But unfortunately none regarding the native mode because we have not figured out a good way to test it automatically.

**Comment by ed2050** [comment]
# Update
I installed nicegui on a fresh mac system with pip.  Browser mode works fine.  But native mode crashes without even opening a window.  Thankfully it gave a helpful error message to console, see below.  This time I'm running exact code from nicegui [styling example](https://github.com/zauberzeug/nicegui/discussions/port=native_mode.find_open_port()) with one change: add `native = True` to `ui.run ()`.

The first error is caused by fork.  I have experience with this.  Unfortunately python fork on macos does bad things.  For instance see [this thread in cpython project](https://github.com/python/cpython/issues/74570#issuecomment-1093748531).  I found [two fixes for fork issues](https://stackoverflow.com/questions/50168647/multiprocessing-causes-python-to-crash-and-gives-an-error-may-have-been-in-progr/72575205#72575205) that may help, but neither works in 100% of cases.  Because of these issues, the multiprocessing package switched from fork to spawn on macos as default process start method [as of python 3.8](https://docs.python.org/3/library/multiprocessing.html?highlight=macOS#contexts-and-start-methods) (see note after start methods or [discussion here](https://github.com/python/cpython/issues/77906)).

# Triggers
Does nicegui issue any forks?  I found the word fork several times in the github code, but no calls to os.fork that I could see.

It can also be triggered by starting processes with mp.  I see native_mode.py creates a new process on lines 107-110:
```
    mp.freeze_support()
    args = host, port, title, width, height, fullscreen, frameless, native.method_queue, native.response_queue
    process = mp.Process(target=open_window, args=args, daemon=False)
    process.start()
```

I don't see an mp context being used, and I can't find mp.set_start_method called elsewhere in nicegui code.  However perhaps another lib nicegui depends on calls set_start_method ('fork').  This would change the start method globally, affecting nicegui mp processes.  I also saw use of mp.Queue and mp.Manager in nicegui, which can create processes that trigger fork bugs.  It's always a good idea to [use an explicit context](https://docs.python.org/3/library/multiprocessing.html?highlight=macOS#multiprocessing.get_context) with multiprocessing.

Third, the bug could be triggered by libraries nicegui uses that call os.fork () directly.  I don't feel up to searching all that code. :)  But we can probably limit it to libs nicegui uses that pywebview doesn't, since pywebview works fine on my machine.

# Error message
There's another error about bad file descriptor that I don't understand the code enough to decipher.  Anyway that seems secondary, triggered during shutdown.  Perhaps it makes sense to someone else.
```
grimey> python3 ngtest.py 
NiceGUI ready to go on http://localhost:8000/
objc[4195]: +[NSString initialize] may have been in progress in another thread when fork() was called.
objc[4195]: +[NSString initialize] may have been in progress in another thread when fork() was called. We cannot safely call it or ignore it in the fork() child process. Crashing instead. Set a breakpoint on objc_initializeAfterForkError to debug.
ERROR:    Traceback (most recent call last):
  File "/Users/ed/Library/Python/3.7/lib/python/site-packages/starlette/routing.py", line 686, in lifespan
    await receive()
  File "/Users/ed/Library/Python/3.7/lib/python/site-packages/starlette/routing.py", line 569, in __aexit__
    await self._router.shutdown()
  File "/Users/ed/Library/Python/3.7/lib/python/site-packages/starlette/routing.py", line 664, in shutdown
    await handler()
  File "/Users/ed/Library/Python/3.7/lib/python/site-packages/nicegui/nicegui.py", line 109, in handle_shutdown
    app.native.main_window.signal_server_shutdown()
  File "/Users/ed/Library/Python/3.7/lib/python/site-packages/nicegui/native.py", line 128, in signal_server_shutdown
    self._send()
  File "/Users/ed/Library/Python/3.7/lib/python/site-packages/nicegui/native.py", line 115, in _send
    method_queue.put((name, args, kwargs))
  File "/usr/local/Cellar/python/3.7.7/Frameworks/Python.framework/Versions/3.7/lib/python3.7/multiprocessing/queues.py", line 82, in put
    if not self._sem.acquire(block, timeout):
OSError: [Errno 9] Bad file descriptor

ERROR:    Application shutdown failed. Exiting.
```

**Reply by rodja** [reply]
Auto-Reload will fork the process. Try `ui.run(reload=False, native=True)`. Also pywebview wants it's own main thread so we start it in a separate process. We are not experiencing any native issues on Mac. There may be some strange configuration causing your errors.

**Reply by ed2050** [reply]
Tried reload=false on both machines, no difference.  Second machine still produces the NSString fork error, which is definitely a mac os issue.  The identifier objc_initializeAfterForkError is a dead giveaway.  But that doesn't tell us what triggers it.

There may be strange config on my machine.  Or there may be strange config on your test machines.  Or there may be problems with particular combinations of versions of macos, nicegui, pywebview, etc, that you haven't encountered in testing.  Who can say?  So hard to diagnose without a debug trace.

I tested on a fresh mac, so something is definitely going on.  What versions of macos, nicegui, pywebview, and cpython have you tested with?

I'm using Pywebview 4.3.3 released earlier this month.  Python I tried 3.8.2 mac native build and 3.8.16 port build on first machine, made no difference.  Second machine python 3.9.6 native build and 3.7.7. port build, again no difference.

**Reply by rodja** [reply]
I'm testing on Pywebview 4.3.3 and Python 3.11. Does http://localhost:8000/ open in the browser correctly? 

Another idea: Maybe you could clone the repo and stepwise remove NiceGUI code until the error disappears.

**Reply by ed2050** [reply]
Thanks, had the same thought.  localhost does work, shows the correct style page even when running in native mode.  So the server is ok, just the native window doesn't display for some reason.

I'll try to track down further when I have time.  Thanks for the suggestion.

**Comment by Muyv** [comment]
check your system proxy
